### PR TITLE
Enable mtls communication with mqtt broker

### PR DESCRIPTION
**Pull request by mteodor**

### CODE REVIEW

File: cmd/main.go [hunk 1]
```diff
@@ -160,38 +184,102 @@ func loadConfig(logger logger.Logger) (config.Config, error) {
...
+		opts.SetTLSConfig(cfg)
+		opts.SetProtocolVersion(4)
 	}
+	client := mqtt.NewClient(opts)
+	token := client.Connect()
+	token.Wait()
 
-	return client
+	if token.Error() != nil {
+		//e := errors.New(fmt.Sprintf("Client %s had error connecting to the broker", name))
```
Comment (drasko) [inline cmd/main.go]: Dead code?

File: cmd/main.go [hunk 2]
```diff
@@ -160,38 +184,102 @@ func loadConfig(logger logger.Logger) (config.Config, error) {
...
 	topic := fmt.Sprintf("channels/%s/messages", ctrlChan)
 	if err := broker.Subscribe(topic); err != nil {
 		logger.Error(fmt.Sprintf("Failed to subscribe to MQTT broker: %s", err.Error()))
 		os.Exit(1)
 	}
 	logger.Info("Subscribed to MQTT broker")
 }
+
+func loadCertificate(cfg *config.MQTTConf) error {
+
```
Comment (drasko) [inline cmd/main.go]: Kill this white line

File: cmd/main.go [hunk 3]
```diff
@@ -160,38 +184,102 @@ func loadConfig(logger logger.Logger) (config.Config, error) {
...
+		pk, _ := ioutil.ReadAll((privKey))
+
+		cert, err = tls.X509KeyPair([]byte(cc), []byte(pk))
+		if err != nil {
+			return err
+		}
+
+		cfg.Cert = cert
+		cfg.CA = caByte
+
```
Comment (drasko) [inline cmd/main.go]: Kill this white line

File: internal/app/agent/api/common.go [hunk 4]
```diff
@@ -26,13 +21,21 @@ type logConf struct {
 }
 
 type mqttConf struct {
-	url string `json:"url"`
+	url         string `json:"url"`
+	username    string `json:"username"`
+	password    string `json:"json"`
+	mtls        bool   `json:"mtls" `
```
Comment (manuio) [inline internal/app/agent/api/common.go]: Rm this white space

File: cmd/main.go [hunk 5]
```diff
@@ -160,38 +184,94 @@ func loadConfig(logger logger.Logger) (config.Config, error) {
...
+func loadCertificate(cfg *config.MQTTConf) error {
+	caByte := []byte{}
+	cert := tls.Certificate{}
+	if cfg.MTLS {
+		caFile, err := os.Open(cfg.CAPath)
+		defer caFile.Close()
+		if err != nil {
+			return err
+		}
+		caByte, _ = ioutil.ReadAll(caFile)
```
Comment (manuio) [inline cmd/main.go]: Don't ignore errors

File: cmd/main.go [hunk 6]
```diff
@@ -75,7 +96,10 @@ func main() {
 	}
 	defer nc.Close()
 
-	mqttClient := connectToMQTTBroker(cfg.Agent.MQTT.URL, cfg.Agent.Thing.ID, cfg.Agent.Thing.Key, logger)
+	mqttClient, err := connectToMQTTBroker(cfg.Agent.MQTT, logger)
+	if err != nil {
+		log.Fatalf(err.Error())
```
Comment (nmarcetic) [inline cmd/main.go]: You should use our logger, you have it imported already and error package for error wrapping.

File: internal/app/agent/api/common.go [hunk 7]
```diff
@@ -26,13 +21,21 @@ type logConf struct {
...
 
 type mqttConf struct {
-	url string `json:"url"`
+	url         string `json:"url"`
+	username    string `json:"username"`
+	password    string `json:"json"`
+	mtls        bool   `json:"mtls"`
+	skipTLSVer  bool   `json:"skip_tls_ver"`
+	retain      bool   `json:"retain"`
+	qoS         int    `json:"qos"`
```
Comment (nmarcetic) [inline internal/app/agent/api/common.go]: Either QoS or qos not qoS.

File: internal/pkg/conn/conn.go [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+	"github.com/mainflux/agent/internal/app/agent"
+	"github.com/mainflux/mainflux/logger"
+	"github.com/mainflux/senml"
+	"github.com/nats-io/go-nats"
+	"robpike.io/filter"
+
+	mqtt "github.com/eclipse/paho.mqtt.golang"
+)
+
+type cmdType string
```
Comment (manuio) [inline internal/pkg/conn/conn.go]: no need to create a type for string

File: cmd/main.go [hunk 9]
```diff
@@ -160,38 +188,97 @@ func loadConfig(logger logger.Logger) (config.Config, error) {
...
+	broker := conn.NewBroker(svc, mc, nc, logger)
 	topic := fmt.Sprintf("channels/%s/messages", ctrlChan)
 	if err := broker.Subscribe(topic); err != nil {
 		logger.Error(fmt.Sprintf("Failed to subscribe to MQTT broker: %s", err.Error()))
 		os.Exit(1)
 	}
 	logger.Info("Subscribed to MQTT broker")
 }
+
+func loadCertificate(cfg *config.MQTTConf) error {
```
Comment (manuio) [inline cmd/main.go]: pointer?

File: cmd/main.go [hunk 10]
```diff
@@ -160,38 +188,97 @@ func loadConfig(logger logger.Logger) (config.Config, error) {
...
+	caByte := []byte{}
+	cert := tls.Certificate{}
+	if cfg.MTLS {
```
Comment (manuio) [inline cmd/main.go]: Maybe better `if !cfg.MTLS {return nil;}`

File: internal/pkg/conn/conn.go [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+		if err := b.svc.Control(uuid, cmdStr); err != nil {
+			b.logger.Warn(fmt.Sprintf("Control operation failed: %s", err))
+		}
+	case exec:
+		b.logger.Info(fmt.Sprintf("Execute command for uuid %s and command string %s", uuid, cmdStr))
+		if _, err := b.svc.Execute(uuid, cmdStr); err != nil {
+			b.logger.Warn(fmt.Sprintf("Execute operation failed: %s", err))
+		}
+	case config:
+		b.logger.Info(fmt.Sprintf("Execute command for uuid %s and command string %s", uuid, cmdStr))
```
Comment (manuio) [inline internal/pkg/conn/conn.go]: Fix `Execute`

File: internal/pkg/conn/conn.go [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+		if err := b.svc.ServiceConfig(uuid, cmdStr); err != nil {
+			b.logger.Warn(fmt.Sprintf("Execute operation failed: %s", err))
```
Comment (manuio) [inline internal/pkg/conn/conn.go]: Same here

File: internal/pkg/conn/conn.go [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+// handleNatsMsg triggered when new message is received on MQTT broker
+func (b *broker) handleNatsMsg(mc mqtt.Client, msg mqtt.Message) {
+	if topic := extractNatsTopic(msg.Topic()); topic != "" {
+		b.nats.Publish(topic, msg.Payload())
+	}
+}
+
+func extractNatsTopic(topic string) string {
+	isEmpty := func(s string) bool {
+		return (len(s) == 0)
```
Comment (manuio) [inline internal/pkg/conn/conn.go]: what is this woodoo?

**Pull request merged**